[Merged by Bors] - chore: remove unnecessary term or tactic `set_option in` with backwards-compatibility flags

grunweg opened the pull request:
As far as I can tell, this removes all `set_option ... false in norm_num` :tada:

Inspired by the linter in #13653 (but not covered by it yet; so far, that linter only checks the set_option *command*).

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg commented:
!bench

grunweg commented on Mathlib/RingTheory/Polynomial/ScaleRoots.lean:
```diff
@@ -170,7 +170,6 @@ theorem scaleRoots_aeval_eq_zero [Algebra R A] {p : R[X]} {a : A} {r : R} (ha :
     (hf : Function.Injective f) {r s : S} (hr : eval₂ f (f r / f s) p = 0)
     (hs : s ∈ nonZeroDivisors S) : eval₂ f (f r) (scaleRoots p s) = 0 := by
-  set_option tactic.skipAssignedInstances false in
   nontriviality S using Subsingleton.eq_zero
```
```suggestion
  -- The proof works without this option, but *much* slower.
  set_option tactic.skipAssignedInstances false in
  nontriviality S using Subsingleton.eq_zero

grunweg commented:
Lesson learned: these options can be performance-sensitive. Reverted the `ScaleRoots` change and added a comment.
@MichaelStollBayreuth Can I ask you to run your script on the benchmarking run above? I'd like to know if there are any other files with significant regressions. Thank you (that script is really nice - I wish the speed-center also allowed sorting by the absolute change)!

MichaelStollBayreuth commented:
```text
General information:
  build:                                               +19.345 * 10⁹ (+0.0159%)

5 files got slower by at least 10⁹ instructions:
  Mathlib.RingTheory.Polynomial.ScaleRoots:            +11.369 * 10⁹ (+49.6%)
  Mathlib.Analysis.SpecialFunctions.Trigonometric.Bounds:
                                                       +4.1656 * 10⁹ (+6.95%)
  Mathlib.Topology.Homotopy.Path:                      +2.7929 * 10⁹ (+3.84%)
  Mathlib.Data.Finset.Pointwise:                       +1.8385 * 10⁹ (+1.25%)
  Mathlib.RingTheory.Localization.Basic:               +1.4865 * 10⁹ (+1.33%)

1 file got slower by at least 10%:
  Mathlib.RingTheory.Polynomial.ScaleRoots:                           +49.6%

2 files got faster by at least 10⁹ instructions:
  Mathlib.Analysis.NormedSpace.PiTensorProduct.InjectiveSeminorm:
                                                       -3.9465 * 10⁹ (-1.32%)
  Mathlib.Algebra.Category.ModuleCat.ChangeOfRings:    -1.1021 * 10⁹ (-0.278%)

No file got faster by at least 10%.
```

grunweg commented:
I looked at the files which slowed down by more than 10^9 instructions. `ScaleRoots` clearly needs to be reverted (with a comment). The simp call in `Data.Finset.Pointwise` speeds up a lot with that option (250 ->50ms locally), hence I also reverted that. In `Trigonometric/Bounds`, the situation seems murkier (the refine's are slow either way, but perhaps go from 850->800ms)... but I've reverted to try another benchmarking run.
The other changes seems inconclusive or a wash; I tend to leave those in.

So, let's re-benchmark this!

grunweg commented:
!bench

MichaelStollBayreuth commented:
<b>General information:</b><tt>
 build:                                               +1.0814 * 10⁹ (+0.000889%)
</tt>

<b>1 file got slower by at least 10⁹ instructions:</b>
<tt> Mathlib.Topology.Homotopy.Path:                      +2.7944 * 10⁹ (+3.84%)
</tt>

<b>No file got slower by at least 10%.</b>

<b>3 files got faster by at least 10⁹ instructions:</b>
<tt> Mathlib.Analysis.NormedSpace.PiTensorProduct.InjectiveSeminorm:
                                                       -2.1722 * 10⁹ (-0.726%)
  Mathlib.Analysis.SpecialFunctions.Trigonometric.Bounds:
                                                       -1.9068 * 10⁹ (-3.18%)
  Mathlib.Algebra.Category.ModuleCat.ChangeOfRings:    -1.1096 * 10⁹ (-0.280%)
</tt>

<b>No file got faster by at least 10%.</b>

semorrison commented:
bors merge

semorrison commented:
bors merge
